### ISSUE TITLE
Linear artifacts in topocalc.horizon horizon function.

### ISSUE THREAD

**Issue by EvanRosenliebNREL** [USDA-ARS-NWRC/topocalc#8]
* topocalc version: 0.3.2
* Python version: 3.7.6
* Operating System: Mac  OS

### Description

Testing the horizon function as a potential tool to help model PV direct beam shading. A 1 by 1 degree SRTM 30 meter tile was loaded in the vicinity of Yosemite. The DEM is projected in CONUS equal area conus projection. The horizon function was tested at multiple azimuth angles, and each exhibited curious linear artifacts that appear to be erroneous. I thought it may have to do with the array transposing, but the artifacts still appear with an azimuth of -90, 0, or 90 which I believe should not require any transposition. 

Examples are shown for azimuth zero, however results are similar with other azimuths with the respective angles (i.e. for azimuth of -90, linear features run east-west instead of north south)

Zoomed in version here. Yellow is high zenith to horizon angle (90 degrees), blue is low zenith to horizon angle:

<img width="662" alt="image" src="https://user-images.githubusercontent.com/58784383/108896110-b62dc680-75d1-11eb-8713-528f61929ad6.png">

A somewhat more zoomed out area -- as you can see it is somewhat common, not a few isolated incidences.

<img width="665" alt="image" src="https://user-images.githubusercontent.com/58784383/108896221-d9f10c80-75d1-11eb-8fe4-38973bcbd6f2.png">

### What I Did

```
import numpy as np
import rasterio as rio
from topocalc.horizon import horizon

ras = rio.open('test_proj.tif')
ar = ras.read(1)
#replace nodata values around edges
ar[ar == -32768] = np.nan

horizon_test_azimuth_0 = np.rad2deg(np.arccos(horizon(0, ar, 30)))

```

**Comment by scotthavens** [comment]
I have seen these before, although these are particularly more noticeable. From you're zoomed in image, it appears that most of the artifacts are coming from the small features in the middle of the larger flat areas. I believe this is coming from the way the horizon is being found along the path. In the `hor1d` function [here](https://github.com/USDA-ARS-NWRC/topocalc/blob/095bc44baaea6b11c852076d3e4cf5fc9cde6a40/topocalc/core_c/hor1d.c#L107), it skips the adjacent pixel to reduce the noise. So it might be fluttering between the actual horizon and a point closer. When it does that, all other points along the line will switch between the actual horizon and artificial one.

Would you mind attaching your test dem? I can see if looking at the adjacent pixel will help or if we'd have to look down the full length of the column.

**Comment by EvanRosenliebNREL** [comment]
Scott, 

Thank you so much for your prompt response.

The original file is far too large, I am attaching an area windowed raster -- I don't believe it is actually the same location as the screenshots above but there are plenty of examples, at least on my end, within this raster. 

[south_horizon_windowed_int.tif.zip](https://github.com/USDA-ARS-NWRC/topocalc/files/6032761/south_horizon_windowed_int.tif.zip)

**Comment by scotthavens** [comment]
Looks like the file is the horizon results? Could you provide the `test_proj.tif` or a subset of it? You should be able to upload 50MB file to this issue or I can provide you a place to put the raster.

**Comment by EvanRosenliebNREL** [comment]
My goodness, I simply drag and dropped the wrong raster there. My apologies!
[test_proj_windowed_int.tif.zip](https://github.com/USDA-ARS-NWRC/topocalc/files/6037827/test_proj_windowed_int.tif.zip)

**Comment by scotthavens** [comment]
No worries, I can replicate what you were seeing so I'll look into it, thanks for reporting!

**Comment by scotthavens** [comment]
It appears that the C code for `horizon` has troubles when the horizon is far way (like a ridge on the other side of a valley) or in relatively flat terrain. This is caused by the `hor1d` function saving time and not calculating the horizon to all points. The solution was to convert the function to pure python and calculate the slope to every point along a 1D horizon. The downside is that this is a bit slower when it's moved to 2D. That will be the next thing to work on.

This image shows a 1D slice from the above grid with the "flutter" quite visible (horizon). The blue is the new horizon calculation which is much smoother in those area.

![image](https://user-images.githubusercontent.com/7975741/110690385-1bbcae00-81a1-11eb-8421-97d3eb55e5d4.png)

**Comment by scotthavens** [comment]
The issue has been resolved in `v0.4.0`. We are currently experiencing problems with Travis CI and don't have a solution for how to publish a new MacOSX wheel to Pypi.

**Comment by scotthavens** [comment]
`v0.5.0` now on PyPi and works for both Linux and OSX.
